File: packages/api-types/src/ws/dto/conversation.ts
export type ConversationCreateRequest = {
  workspace_id?: string | null;
  project_id?: string | null;
  cwd?: string | null;
  provider_id: string;
  model?: string | null;
  thinking?: string | null;
  title?: string | null;
};

export type ConversationListRequest = {
  workspace_id?: string | null;
  project_id?: string | null;
  cwd?: string | null;
  cursor?: string | null;
  limit?: number | null;
};

export type ConversationIdRequest = {
  conversation_id: string;
};

export type ConversationMessagesRequest = {
  conversation_id: string;
  before_seq?: number | null;
  limit?: number | null;
};

export type ConversationRenameRequest = {
  conversation_id: string;
  title: string;
};

export type ConversationConfigureRequest = {
  conversation_id: string;
  provider_id?: string | null;
  model?: string | null;
  thinking?: string | null;
};

export type ConversationSubscribeRequest = {
  conversation_id: string;
  after_sequence?: number | null;
};

export type ConversationSendRequest = {
  conversation_id: string;
  text: string;
  attachment_paths?: string[] | null;
};

export type ConversationSteerRequest = {
  conversation_id: string;
  expected_turn_id: string;
  text: string;
};

export type ConversationQueueAddRequest = {
  conversation_id: string;
  text: string;
  attachment_paths?: string[] | null;
};

export type ConversationQueueUpdateRequest = {
  conversation_id: string;
  item_id: string;
  text?: string | null;
  status?: string | null;
};

export type ConversationQueueReorderRequest = {
  conversation_id: string;
  item_ids: string[];
};

export type ConversationQueueDeleteRequest = {
  conversation_id: string;
  item_id: string;
};

export type ConversationPermissionRespondRequest = {
  conversation_id: string;
  request_id: string;
  option_id?: string | null;
  allowed?: boolean | null;
};

export type AgentModelCatalogGetRequest = {
  agent_id: string;
  refresh?: boolean | null;
};

export type ConversationMeta = {
  id: string;
  created_at: string;
  updated_at: string;
  deleted: boolean;
  title: string | null;
  cwd: string;
  workspace_id: string | null;
  project_id: string | null;
  provider_id: string;
  last_message_at: string | null;
  last_event_seq: number;
  persistence_handle: string | null;
  runtime_status: string;
  selected_model: string | null;
  selected_thinking: string | null;
  selected_mode: string | null;
  supports_steer: boolean;
};

export type ConversationIndexEntry = {
  id: string;
  title: string | null;
  cwd: string;
  workspace_id: string | null;
  project_id: string | null;
  provider_id: string;
  updated_at: string;
  last_message_at: string | null;
  deleted: boolean;
};

export type ConversationListResponse = {
  items: ConversationIndexEntry[];
};

export type ConversationTurnIdResponse = {
  turn_id: string;
};

export type ConversationSubscribeResponse = {
  last_event_seq: number;
};

export type ConversationClientEvent = {
  conversation_id: string;
  event_id: string;
  sequence: number;
  payload: Record<string, unknown> & { type: string };
};

export type AgentModelCatalog = {
  agent_id: string;
  status: "ok" | "unsupported" | "auth_required" | "error" | "probing";
  models: Array<{
    id: string;
    label: string;
    group?: string | null;
    is_default?: boolean;
    thinking?: unknown;
  }>;
  modes: Array<{ id: string; label: string; is_default?: boolean }>;
  thinking: unknown;
  strategies_used: string[];
  fetched_at: string;
  source: "cache" | "live";
  message: string | null;
};

export type AgentModelCatalogUpdated = {
  agent_id: string;
  catalog: AgentModelCatalog;
};

File: apps/api/src/api/ws/message/conversation.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ConversationCreateRequest {
    #[serde(default)]
    pub workspace_id: Option<String>,
    #[serde(default)]
    pub project_id: Option<String>,
    #[serde(default)]
    pub cwd: Option<String>,
    pub provider_id: String,
    #[serde(default)]
    pub model: Option<String>,
    #[serde(default)]
    pub thinking: Option<String>,
    #[serde(default)]
    pub title: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ConversationListRequest {
    #[serde(default)]
    pub workspace_id: Option<String>,
    #[serde(default)]
    pub project_id: Option<String>,
    #[serde(default)]
    pub cwd: Option<String>,
    #[serde(default)]
    pub cursor: Option<String>,
    #[serde(default)]
    pub limit: Option<u32>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ConversationIdRequest {
    pub conversation_id: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ConversationMessagesRequest {
    pub conversation_id: String,
    #[serde(default)]
    pub before_seq: Option<u64>,
    #[serde(default)]
    pub limit: Option<u32>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ConversationRenameRequest {
    pub conversation_id: String,
    pub title: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ConversationSubscribeRequest {
    pub conversation_id: String,
    #[serde(default)]
    pub after_sequence: Option<u64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ConversationSendRequest {
    pub conversation_id: String,
    pub text: String,
    #[serde(default)]
    pub attachment_paths: Option<Vec<String>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ConversationSteerRequest {
    pub conversation_id: String,
    pub expected_turn_id: String,
    pub text: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ConversationQueueAddRequest {
    pub conversation_id: String,
    pub text: String,
    #[serde(default)]
    pub attachment_paths: Option<Vec<String>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ConversationQueueUpdateRequest {
    pub conversation_id: String,
    pub item_id: String,
    #[serde(default)]
    pub text: Option<String>,
    #[serde(default)]
    pub status: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ConversationQueueReorderRequest {
    pub conversation_id: String,
    pub item_ids: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ConversationQueueDeleteRequest {
    pub conversation_id: String,
    pub item_id: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ConversationPermissionRespondRequest {
    pub conversation_id: String,
    pub request_id: String,
    #[serde(default)]
    pub option_id: Option<String>,
    #[serde(default)]
    pub allowed: Option<bool>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentModelCatalogGetRequest {
    pub agent_id: String,
    #[serde(default)]
    pub refresh: Option<bool>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ConversationOk {
    pub ok: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ConversationConfigureRequest {
    pub conversation_id: String,
    #[serde(default)]
    pub provider_id: Option<String>,
    #[serde(default)]
    pub model: Option<String>,
    #[serde(default)]
    pub thinking: Option<String>,
}
